### DISCUSSION TITLE
default value for Auto Properties of struct

### DISCUSSION THREAD

**Discussion by MkazemAkhgary** [dotnet/csharplang#1083] [General]
based on this conclusion #1082, having default constructor for structs will make a disagreement between developers whether `default` should have same value as `new S` or not.

This proposal is trying to avoid this ambiguity by solving problem in different way. 

before we start, I want to mention that `default` will be always zero and default constructor is not allowed. same as before. now here is proposal:

allow to set default value for auto properties of an struct.

---

**Auto Properties default (applicable with constants):**

```
public struct S
{
    // assume there are some constructors here...
    public int Foo { get; private set; } = 10;
}
```

Equivalent of:

```
public struct S
{
    public int Foo 
    {       
       int val;
       const int def = 10;
       get => unchecked((int)(val + def));
       set => unchecked(value = (int)(val - def));
    }
}
```

As you can see in this workaround, we used offset value to achieve practical default value of struct. it doesn't matter if you use `new S` or `default(S)`, default value will be always what you **explicitly** set to property.

notes: 
 - cast is important, in case of using unsigned integer types or bytes or enums.
 - unchecked is important to prevent possible exception at runtime.

benefits:

 -  you can set default value of struct's property, that means you have default value always as expected
 - no overhead in terms of allocating memory. its same as before, just a zeroed memory.
 - struct size does not change.

drawbacks: 
 - the obvious overhead is addition and subtraction and cast, but that's very negligible.
 - the overflow in add/sub requires unchecked statement (but doesn't cause incorrect result).
 - this approach is only applicable for limited set of types (that have numeric semantic).

rules for this section:
 - default must be evaluated at compile time. (issue #1028 helps dealing with this limitation)
 - property type must be either of this list

> byte, sbyte, ushort, short, uint, int, ulong, long, char, float, double, decimal or any type of Enum.

---

**Auto Properties default (for Boolean):**

for Booleans this is achieved simpler.

```
public struct S
{
    public bool Foo { get; private set; } = true;
}
```

Equivalent of:

```
public struct S
{
    public bool Foo 
    {
       bool val;
       const bool def = true;
       get => def ? !val : val;
       set =>  val = def ? !value : value;
    }
}
```

---

**Additional support for String, Reference types, Generic types (applicable with lazy evaluation):**

Similar to singleton pattern, use a flag.

```
public struct S
{
    public string Foo { get; private set; } = GetResult(); // or "Hello World" or anything valid
}
```

Equivalent of:

```
public struct S
{
    public string Foo 
    {
       string? val;
       get => val ?? (val = GetResult()).Value;
       set => val = value;
    }
}
```

notes:
 - here `string?` was used for clarity, (valid syntax from #790)
 - if string literal is used as default value, constant string value could be used.

benefits:
 - there you can have your default value for structs, like the way you can have with classes.

drawbacks of this section:
 - struct size increases (potentially more than a byte) and therefore it may not be expected by developer.
---

**Comment by jnm2** [comment]
I've asked before and I'd prefer this: can we do the right thing and make the CLR pattern-fill instead of zero-init for structs that have fields initialized with constant data? You can do this at great speed.
